Google.Api.Gax.Grpc/HeaderParameterExtractor.cs

```diff
@@ -13,54 +13,89 @@
 namespace Google.Api.Gax.Grpc
 {
     /// <summary>
-    /// 
+    /// Collects the explicit routing header extraction instructions and
+    /// extracts the routing header parameters from a specific request
+    /// using these instructions.
+    /// This class is immutable.
     /// </summary>
     public sealed class HeaderParameterExtractor<TRequest>
     {
         private readonly IReadOnlyList<ParamExtraction> _extractions;
+        private readonly IReadOnlyDictionary<string, IEnumerable<ParamExtraction>> _groupings;
 
         /// <summary>
-        /// 
+        /// Create a new HeaderParameterExtractor
         /// </summary>
-        /// <param name="paramName"></param>
-        /// <param name="regex"></param>
-        /// <param name="selector"></param>
-        public HeaderParameterExtractor(string paramName, Regex regex, Func<TRequest, string> selector)
+        public HeaderParameterExtractor()
         {
-            _extractions = new List<ParamExtraction> { new ParamExtraction(paramName, regex, selector) };
+            _extractions = new List<ParamExtraction>();
         }
 
         private HeaderParameterExtractor(IEnumerable<ParamExtraction> extractions)
         {
             _extractions = new List<ParamExtraction>(extractions);
+
+            // It is important that we Reverse the parameters within the group.
+            // The extractions follow the `last successfully matched wins` rule for
+            // conflict resolution when there are multiple extractions for the same parameter name.
+            // (see `google/api/routing.proto` for further details)
+            // But we would prefer to stop evaluating on the first successfully matched extraction,
+            // thus the reverse.
+            _groupings = _extractions.GroupBy(pe => pe.ParamName)
+                .ToDictionary(paramGroup => paramGroup.Key,
```

Comment:
Nit - I'd unwrap this last method call to a single line. I'm happy to do that later though.